storage/node_liveness_test.go

```diff
@@ -0,0 +1,172 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Spencer Kimball (spencer@cockroachlabs.com)
+
+package storage_test
+
+import (
+	"testing"
+
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+
+	"github.com/cockroachdb/cockroach/gossip"
+	"github.com/cockroachdb/cockroach/gossip/resolver"
+	"github.com/cockroachdb/cockroach/roachpb"
+	"github.com/cockroachdb/cockroach/rpc"
+	"github.com/cockroachdb/cockroach/storage"
+	"github.com/cockroachdb/cockroach/util"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+	"github.com/cockroachdb/cockroach/util/metric"
+	"github.com/cockroachdb/cockroach/util/netutil"
+)
+
+func verifyLiveness(t *testing.T, mtc *multiTestContext) {
+	util.SucceedsSoon(t, func() error {
+		for _, nl := range mtc.nodeLivenesses {
+			for _, g := range mtc.gossips {
+				live, err := nl.IsLive(g.GetNodeID())
+				if !live {
+					return errors.Errorf("node %d not live", g.GetNodeID())
+				} else if err != nil {
+					return err
+				}
+			}
+		}
+		return nil
+	})
+}
+
+func TestNodeLiveness(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	mtc := startMultiTestContext(t, 3)
+	defer mtc.Stop()
+
+	// Verify liveness of all nodes for all nodes.
+	verifyLiveness(t, mtc)
+
+	// Advance clock past the liveness threshold to verify IsLive becomes false.
+	mtc.manualClock.Increment(storage.LivenessThreshold.Nanoseconds() + 1)
+	for idx, nl := range mtc.nodeLivenesses {
+		nodeID := mtc.gossips[idx].GetNodeID()
+		live, err := nl.IsLive(nodeID)
+		if live {
+			t.Errorf("expected node %d to be considered not-live after advancing node clock", nodeID)
+		} else if err != nil {
+			t.Error(err)
+		}
+	}
+	// Trigger a manual heartbeat and verify liveness is reestablished.
+	for _, nl := range mtc.nodeLivenesses {
+		nl.ManualHeartbeat()
```

Comment:
This test would pass even without the manual heartbeat because the heartbeat ticker uses real time and would trigger a new heartbeat within the SucceedsSoon timeout. You need to disable the ticker or set it to something impossibly long for this test.